tsdb/engine/tsm1/engine.go

```diff
@@ -1193,7 +1195,7 @@ func (e *Engine) overlay(r io.Reader, basePath string, asNew bool) error {
 			return err
 		}
 	}
-	return nil
```

Comment:
This usage of `WriteToFileNoLock()` is pretty tricky - essentially I think we're relying on the `Shard` mutex several stack frames above to ensure there are no concurrent writes. Is there any way to make this more obviously safe / less prone to future breakage?

For example, we could have a separate read/write mutex for messing with `fields.idx`.

If we can't make this more obviously safe I think we need some extensive comments explaining the intended locking behaviour.

Reply:
We can lock on the MeasurementFieldSet, I believe. I will verify that.

Reply:
Trying lock/unlock in `marshalMeasurementFieldSet`

Reply:
We're locking not to avoid concurrent writes (because those are okay, they will synchronize on the file rename), but to ensure unsullied access to the `MeasurementFieldSet` while marshaling.

Reply:
Reworked locks to cover a broader set of operations in writing the file out.